rfcs/function-debug-id.md

```diff
@@ -0,0 +1,27 @@
+# `debug.id`
+
+## Summary
+
+Add `debug.id` which accepts a value and always returns the identifier unique to that value.
+
+## Motivation
+
+When debugging, it is often useful to compare identity. Usually `print` will print the memory address of the table or userdata, but when `__tostring` is overriden, the address becomes unobtainable.
+
+`rawequal` can compare equality between two arguments, but is not suitable for a human comparing output between two `print` calls or between keys or values in a table.
+
+## Design
+
+`debug.id(value: string | userdata | table | thread | function): number` always returns the identifier unique to that value, and is not possible to override. The identifier can be reused after the value associated with it is garbage collected.
+
+While tables and userdata have individualized metatables, also allowing functions and threads to be passed to `debug.id` would let their type-level `__tostring` be redefined. Strings are useful to accept because of long, similar strings.
```

Comment:
The mention of `__tostring` is a little confusing here, although I understand what it means after a few tries :) Generally speaking I think we'd want a function like this to accept any value with reference semantics which includes threads/functions regardless of what gets overridden where.

Reply:
I meant it as a justification for including all "reference" types, I'll remove that.

Reply:
I meant that as a justification to include all "reference" types, I'll remove it since it's confusing.

Reply:
> accept any value with reference semantics

I was trying to justify this, I'll remove that language since it's confusing.